<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Species range maps</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Macroecology and global change</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pracs: macroecological analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="a0_setup.html">0. Getting started</a>
    </li>
    <li>
      <a href="a1_SpatialData.html">1. Spatial data in R</a>
    </li>
    <li>
      <a href="a2_RichnessGradients.html">2. Species richness gradients</a>
    </li>
    <li>
      <a href="a3_RichnessRegression.html">3. Species richness regression</a>
    </li>
    <li>
      <a href="a4_RangeMaps.html">4. Species range maps</a>
    </li>
    <li>
      <a href="a5_SpeciesThreats.html">5. Species threats</a>
    </li>
    <li>
      <a href="a6_BiodivChanges.html">6. Analysing biodiversity changes</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pracs: species distribution modelling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="b1_SpeciesData.html">1. Species data</a>
    </li>
    <li>
      <a href="b2_EnvData.html">2. Environmental data</a>
    </li>
    <li>
      <a href="b3_SDM_intro.html">3. SDMs: simple model fitting</a>
    </li>
    <li>
      <a href="b4_SDM_eval.html">4. SDMs: assessment and prediction</a>
    </li>
    <li>
      <a href="b5_pseudoabsence.html">5. Pseudo-absence and background data</a>
    </li>
    <li>
      <a href="b6_SDM_algorithms.html">6. SDMs: algorithms</a>
    </li>
    <li>
      <a href="b7_SDM_ensembles.html">7. SDMs: ensembles</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.uni-potsdam.de/en/ibb-macroecology/index">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/ZurellLab">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Species range maps</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<hr />
<div class="alert alert-info">
<p><strong>RStudio project</strong></p>
<p>Open the RStudio project that we created in the first session. I
recommend to use this RStudio project for the entire course and within
the RStudio project create separate R scripts for each session.</p>
<ul>
<li>Create a new empty R script by going to the tab “File”, select “New
File” and then “R script”</li>
<li>In the new R script, type
<code># Session a4: Species range maps</code> and save the file in your
folder “scripts” within your project folder, e.g. as
“a4_RangeMaps.R”</li>
</ul>
</div>
<p>We are living in an age of big data where biodiversity are becoming
increasingly available in digital format and at a global scale <span
class="citation">(Wüest et al. 2020)</span>. Many different types of
biodiversity data exist, e.g. from standardised monitoring schemes,
citizen science platforms, or expert knowledge. Each of these data types
comes with own challenges. Within this module, I want to provide some
first impressions on how you can obtain and process typical types of
species data. Specifically, in this session, we will work with range
maps of terrestrial animals and plants, while occurrence data for
terrestrial species in <a href="https://www.gbif.org/">GBIF</a> (the
Global Biodiversity Information facility) and for marine species in <a
href="https://obis.org/">OBIS</a> (the Ocean Biodiversity Information
System) will be covered later in the <a
href="b1_SpeciesData.html"><em>species distribution modelling</em>
part</a>.</p>
<div id="obtaining-range-maps" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Obtaining range
maps</h1>
<p>We rarely have detailed biodiversity data available over large
geographic extents. At broad (continental to global) extents,
expert-drawn range maps (also called extent-of-occurrence maps) are
often the primary data source on species distributions.</p>
<div id="iucn-range-maps" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> IUCN range maps</h2>
<p>The IUCN (the International Union for the Conservation of Nature)
provides expert range maps for a large number of species including
mammals, birds (through <a href="http://www.birdlife.org/">BirdLife
International</a>), amphibians, reptiles, and freshwater fishes. There
are also some range maps on plants and marine species, but these are
very limited taxonomically. Have a look for which taxa range maps are
available: <a
href="https://www.iucnredlist.org/resources/spatial-data-download"
class="uri">https://www.iucnredlist.org/resources/spatial-data-download</a>.
You can download them for free, but you should provide some information
on your work to obtain the data.</p>
<p>Most of the IUCN data are provided in the form of shapefiles. In
practical 2, we have already loaded the range map of the <a
href="https://damariszurell.github.io/EEC-Macro/data/IUCN_Sorex_alpinus.zip">Alpine
shrew</a> and learned how to use the package <code>terra</code> for
reading in the shapefiles. You may remember that the shapefile is
recognized as <code>SpatVector</code> object.</p>
<pre class="r"><code>library(terra)

# Load the shapefile
(shrew &lt;- terra::vect(&#39;data/IUCN_Sorex_alpinus.shp&#39;))</code></pre>
<pre><code>##  class       : SpatVector 
##  geometry    : polygons 
##  dimensions  : 1, 27  (geometries, attributes)
##  extent      : 5.733728, 26.67935, 42.20601, 51.89984  (xmin, xmax, ymin, ymax)
##  source      : IUCN_Sorex_alpinus.shp
##  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
##  names       : id_no      binomial presence origin seasonal compiler yrcompiled
##  type        : &lt;chr&gt;         &lt;chr&gt;    &lt;int&gt;  &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;      &lt;int&gt;
##  values      : 29660 Sorex alpinus        1      1        1     IUCN       2008
##         citation source dist_comm (and 17 more)
##            &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;              
##  IUCN (Internat~     NA        NA</code></pre>
<pre class="r"><code># Plot the Central Europe
library(maps)
maps::map(&#39;world&#39;,xlim=c(5,30), ylim=c(40,55))

# Overlay the range of the Alpine Shrew
plot(shrew, col=&#39;red&#39;, add=T)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>Unfortunately, there is no API for the IUCN range maps. So, you need
to register with IUCN and then you can download range maps for different
taxonomic groups. The range maps for birds come as separate shape files
per species, while the range maps of mammals are provided as single
shapefile that contains all species. Course participants can download
the mammal shapefile from the moodle course - of course, the license
agreement by the IUCN apply!</p>
<pre class="r"><code># Read shapefile for all mammals using the raster package:
mammals &lt;- terra::vect(&#39;data/MAMMTERR.shp&#39;)</code></pre>
<p>The shapefile contains the range polygons for all described mammal
species. The attribute table contains information on species’ PRESENCE,
ORIGIN, and SEASONALity. Please have a look at the metadata to
understand the different values these attributes can be coded as.</p>
<pre class="r"><code>mammals</code></pre>
<pre><code>##  class       : SpatVector 
##  geometry    : polygons 
##  dimensions  : 43444, 14  (geometries, attributes)
##  extent      : -180.0001, 180.0025, -56.66595, 90  (xmin, xmax, ymin, ymax)
##  source      : MAMMTERR.shp
##  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
##  names       :        BINOMIAL PRESENCE ORIGIN SEASONAL COMPILER  YEAR
##  type        :           &lt;chr&gt;    &lt;int&gt;  &lt;int&gt;    &lt;int&gt;    &lt;chr&gt; &lt;int&gt;
##  values      : Amblysomus cor~        1      1        1     IUCN  2008
##                Amblysomus hot~        1      1        1     IUCN  2008
##                Amblysomus mar~        1      1        1     IUCN  2008
##         CITATION SOURCE DIST_COMM ISLAND (and 4 more)
##            &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;             
##  IUCN (Internat~     NA        NA     NA             
##  IUCN (Internat~     NA        NA     NA             
##  IUCN (Internat~     NA        NA     NA</code></pre>
<pre class="r"><code># Inspect attribute table
terra::head(mammals)</code></pre>
<pre><code>##                     BINOMIAL PRESENCE ORIGIN SEASONAL COMPILER YEAR
## 1         Amblysomus corriae        1      1        1     IUCN 2008
## 2     Amblysomus hottentotus        1      1        1     IUCN 2008
## 3         Amblysomus marleyi        1      1        1     IUCN 2008
## 4        Amblysomus robustus        1      1        1     IUCN 2008
## 5 Amblysomus septentrionalis        1      1        1     IUCN 2008
## 6 Amblysomus septentrionalis        1      1        1     IUCN 2008
##                                                CITATION SOURCE DIST_COMM ISLAND
## 1 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
## 2 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
## 3 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
## 4 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
## 5 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
## 6 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
##   SUBSPECIES SUBPOP  SHAPE_area SHAPE_len
## 1       &lt;NA&gt;   &lt;NA&gt;  4.93503476 15.575123
## 2       &lt;NA&gt;   &lt;NA&gt; 19.94471634 28.392110
## 3       &lt;NA&gt;   &lt;NA&gt;  0.12779360  1.902574
## 4       &lt;NA&gt;   &lt;NA&gt;  0.09936845  1.655898
## 5       &lt;NA&gt;   &lt;NA&gt;  0.25763915  2.078154
## 6       &lt;NA&gt;   &lt;NA&gt;  2.70333866  7.080348</code></pre>
<p>We can search for specific species or species groups in the attribute
table in different ways:</p>
<pre class="r"><code># Range map for the species &#39;Lynx lynx&#39;
terra::subset(mammals, mammals$BINOMIAL == &quot;Lynx lynx&quot;)

# Show all entries for the species with the word &#39;Lynx&#39; in their name
grep(&#39;Lynx&#39;,mammals$BINOMIAL, value=T)

# Range map for all species with the word &#39;Lynx&#39; in their name
mammals[grep(&#39;Lynx&#39;, mammals$BINOMIAL),]</code></pre>
<p>We can use the attribute table subsets to select specific polygons
that we want to look at.</p>
<pre class="r"><code># Assign range map of the Eurasian lynx to separate object
lynx_lynx &lt;- mammals[mammals$BINOMIAL==&#39;Lynx lynx&#39;,]

# Map the range
maps::map(&#39;world&#39;)
plot(lynx_lynx, col=&#39;red&#39;, add=T)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div class="alert alert-success" role="alert">
<p><strong>Test it yourself</strong></p>
<ul>
<li>Select another mammal species and plot the range map.</li>
</ul>
</div>
</div>
<div id="bien-range-maps" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> BIEN range maps</h2>
<p>The <a href="https://bien.nceas.ucsb.edu/bien/biendata/">BIEN</a>
database (Botanical Information and Ecology Network) contains many range
maps on plants, but unfortunately only for the Americas. These can be
accessed using the <code>BIEN</code> package. As illustrative example,
we load the range map for the monkey puzzle tree (or Chilean pine -
<em>Araucaria araucana</em>).</p>
<pre class="r"><code>library(BIEN)

# Load the range map for the monkey puzzle
(monkey_puzzle_sf &lt;- BIEN_ranges_load_species(&#39;Araucaria_araucana&#39;))</code></pre>
<pre><code>## Simple feature collection with 1 feature and 2 fields
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -109.9819 ymin: -55.73974 xmax: -55.46083 ymax: 62.67197
## Geodetic CRS:  WGS 84
##              species  gid                       geometry
## 1 Araucaria_araucana 6382 MULTIPOLYGON (((-72.94742 -...</code></pre>
<p>The function <code>BIEN_ranges_load_species()</code> will output an
<code>sf</code> polygon, which we simply convert into a
<code>SpatVector</code> object that we already know from the
<code>terra</code> package.</p>
<pre class="r"><code># Change into SpatVector object (terra package)
(monkey_puzzle_terra &lt;- terra::vect(monkey_puzzle_sf))</code></pre>
<pre><code>##  class       : SpatVector 
##  geometry    : polygons 
##  dimensions  : 1, 2  (geometries, attributes)
##  extent      : -109.9819, -55.46083, -55.73974, 62.67197  (xmin, xmax, ymin, ymax)
##  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
##  names       :            species   gid
##  type        :              &lt;chr&gt; &lt;int&gt;
##  values      : Araucaria_araucana  6382</code></pre>
<pre class="r"><code># Map
maps::map(&#39;world&#39;,xlim = c(-180,-20),ylim = c(-80,80))
plot(monkey_puzzle_terra,col=&#39;red&#39;,add=T)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>The native range of the mokey puzzle is in the Chilean Andes, so
clearly the range maps also show areas where the species
naturalized.</p>
</div>
</div>
<div id="working-with-range-maps" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Working with range
maps</h1>
<div id="range-size-and-range-centre" class="section level2"
number="2.1">
<h2><span class="header-section-number">2.1</span> Range size and range
centre</h2>
<p>The <code>terra</code> package allows us to easily calculate the area
of the polygons, meaning the range size of our species. The function
<code>expanse()</code> outputs the area in square meters, kilometers or
hectars.</p>
<pre class="r"><code># Range area of alpine shrew in square kilometers:
terra::expanse(shrew, unit=&quot;km&quot;)</code></pre>
<pre><code>## [1] 490543.3</code></pre>
<pre class="r"><code># Range area of monkey puzzle in square kilometers:
terra::expanse(monkey_puzzle_terra, unit=&quot;km&quot;)</code></pre>
<pre><code>## [1] 5783633</code></pre>
<p>We can also very easily calculate the centre of gravity or range
centroid from the <code>SpatVector</code> object.</p>
<pre class="r"><code># Range centroid:
terra::centroids(shrew)</code></pre>
<pre><code>##  class       : SpatVector 
##  geometry    : points 
##  dimensions  : 1, 27  (geometries, attributes)
##  extent      : 16.02645, 16.02645, 46.90336, 46.90336  (xmin, xmax, ymin, ymax)
##  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
##  names       : id_no      binomial presence origin seasonal compiler yrcompiled
##  type        : &lt;chr&gt;         &lt;chr&gt;    &lt;int&gt;  &lt;int&gt;    &lt;int&gt;    &lt;chr&gt;      &lt;int&gt;
##  values      : 29660 Sorex alpinus        1      1        1     IUCN       2008
##         citation source dist_comm (and 17 more)
##            &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;              
##  IUCN (Internat~     NA        NA</code></pre>
<pre class="r"><code># Map the species range and add the centroid to the map
maps::map(&#39;world&#39;,xlim=c(5,30), ylim=c(40,55))
plot(shrew, col=&#39;red&#39;, add=T)
points(terra::centroids(shrew), col=&#39;blue&#39;,cex=3,pch=&quot;+&quot;)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>We need to be careful how to interpret these centroids. They
represent the centre of gravity, i.e. the mean coordinates of the
distribution (weighted by cell size) but obviously, if we have several
patches, the centroid might not even fall within an occupied patch.</p>
<p>If you have a polygons object with multiple layers, you will receive
a vector of areas and also multiple centroids. In such case, you first
have to dissolve the multiple polygon layers. Let’s look at one example.
In the <code>mammals</code> shapefile, the alpine shrew range maps are
actually multiple layers.</p>
<pre class="r"><code># Filter the mammals shapefiles for the shrew
shrew2 &lt;- mammals[mammals$BINOMIAL==&#39;Sorex alpinus&#39;,]

# Range area of alpine shrew in square kilometers:
terra::expanse(shrew2, unit=&quot;km&quot;) # should produce multiple values</code></pre>
<pre><code>## [1] 179581.6854 310495.2803    466.3444</code></pre>
<pre class="r"><code># Use aggregate() to dissolve the polygon and recalculate area
terra::expanse(aggregate(shrew), unit=&quot;km&quot;)</code></pre>
<pre><code>## [1] 490543.3</code></pre>
<pre class="r"><code># Map the species range and add the centroid to the map
maps::map(&#39;world&#39;,xlim=c(5,30), ylim=c(40,55))
plot(shrew, col=&#39;red&#39;, add=T)
points(terra::centroids(shrew2), col=&#39;blue&#39;,cex=3,pch=&quot;+&quot;)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<pre class="r"><code># Map the species range and add the centroid to the map
maps::map(&#39;world&#39;,xlim=c(5,30), ylim=c(40,55))
plot(shrew, col=&#39;red&#39;, add=T)
points(terra::centroids(aggregate(shrew2)), col=&#39;blue&#39;,cex=3,pch=&quot;+&quot;)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-11-2.png" width="672" /></p>
<p>In case of the Monkey Puzzle, we have suspected that the range maps
also contain non-native areas. We can clip the range maps to a desired
spatial extent and then only calculate the range centroid for this
(presumed) native range.</p>
<pre class="r"><code># Let&#39;s clip the range polygons to (rough coordinates of) South America
monkey_puzzle_SAm &lt;- terra::intersect(monkey_puzzle_terra, ext(-85, -30, -55, 5))

# Map the range and range centroid
maps::map(&#39;world&#39;,xlim = c(-100,-10),ylim = c(-60,15))
plot(monkey_puzzle_SAm,col=&#39;red&#39;,add=T)
points(terra::centroids(monkey_puzzle_SAm), col=&#39;blue&#39;,pch=&#39;+&#39;, cex=3)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<div class="alert alert-success" role="alert">
<p><strong>Test it yourself</strong></p>
<ul>
<li>Use the range map of the mammal species you selected earlier, and
calculate the range size and range centroid. Add the point location of
your centroid to the range map that you have plotted previously.</li>
</ul>
</div>
</div>
<div id="rasterising-range-maps" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Rasterising range
maps</h2>
<p>For many applications in macroecology, we need to rasterise the
polygons. The problem is that it is unclear at which spatial resolution
the range maps accurately represent species occurrences. <span
class="citation">Hurlbert and Jetz (2007)</span> and <span
class="citation">Jetz, McPherson, and Guralnick (2012)</span> define the
minimum spatial resolution as 100-200km (1-2°), although also
resolutions of 50km (0.5°) have been used <span class="citation">(Krosby
et al. 2015; Zurell et al. 2018)</span>.</p>
<div id="rasterising-range-maps-with-terra" class="section level3"
number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Rasterising range
maps with <code>terra</code></h3>
<p>Rasterising polygon data is made very easy in the <code>terra</code>
package. We first have to define a <code>SpatRaster</code> object of the
desired resolution, and then transfer the polygon data to the raster
cells.</p>
<pre class="r"><code># By default, terra() will create a 1° resolution map in the *WGS 84* coordinate system (lon/lat).
(r_1deg &lt;- terra::rast())</code></pre>
<pre><code>## class       : SpatRaster 
## dimensions  : 180, 360, 1  (nrow, ncol, nlyr)
## resolution  : 1, 1  (x, y)
## extent      : -180, 180, -90, 90  (xmin, xmax, ymin, ymax)
## coord. ref. : lon/lat WGS 84</code></pre>
<pre class="r"><code># Now, rasterise the shrew polygon data to the 1° raster grid
(shrew_1deg &lt;- terra::rasterize(shrew, r_1deg))</code></pre>
<pre><code>## class       : SpatRaster 
## dimensions  : 180, 360, 1  (nrow, ncol, nlyr)
## resolution  : 1, 1  (x, y)
## extent      : -180, 180, -90, 90  (xmin, xmax, ymin, ymax)
## coord. ref. : lon/lat WGS 84 
## source(s)   : memory
## name        : layer 
## min value   :     1 
## max value   :     1</code></pre>
<pre class="r"><code>maps::map(&#39;world&#39;,xlim=c(5,30), ylim=c(40,55))
plot(shrew, col=&#39;red&#39;, add=T)
plot(shrew_1deg, add=T, alpha=0.6, legend=F)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Obviously, the margins of the range polgyon and the raster map differ
at several places.</p>
<div class="alert alert-success" role="alert">
<p><strong>Test it yourself</strong></p>
<ul>
<li>Check out the help page <code>?rasterize</code> and find out what
the argument <code>touches</code> is doing. Rasterise the shrew range
map again with setting a different <code>touches</code> argument and map
the result. What is the difference?</li>
<li>What does the <code>cover</code> argument do?</li>
</ul>
</div>
<p>We look at a second example, the lynx:</p>
<pre class="r"><code># Define an empty SpatRaster of the world at 2° spatial resolution
(r_2deg &lt;- terra::rast(res=2))</code></pre>
<pre><code>## class       : SpatRaster 
## dimensions  : 90, 180, 1  (nrow, ncol, nlyr)
## resolution  : 2, 2  (x, y)
## extent      : -180, 180, -90, 90  (xmin, xmax, ymin, ymax)
## coord. ref. : lon/lat WGS 84</code></pre>
<pre class="r"><code># Rasterize the eurasian lynx data
lynx_lynx_2deg &lt;- terra::rasterize(lynx_lynx, r_2deg)

# Map the occupied grid cells
maps::map(&#39;world&#39;)
plot(lynx_lynx_2deg, add = TRUE, legend = FALSE)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
<div id="rasterising-range-maps-with-letsr" class="section level3"
number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> Rasterising range
maps with <code>letsR</code></h3>
<p>There are also specific macroecological packages in R that facilitate
working with range maps and rasterising them, for example the function
<code>lets.presab()</code> in the <code>letsR</code> package.</p>
<p>At the moment, this package still relies on the older spatial R
package <code>raster</code> and is not available for the newest R
version. We thus have to install an older version and need some
workarounds for the older <code>raster</code> standards.</p>
<pre class="r"><code>install_version(&quot;letsR&quot;, version = &quot;4.0&quot;, repos = &quot;http://cran.us.r-project.org&quot;)</code></pre>
<pre class="r"><code>library(letsR)
library(raster)

# The lets.presab() function expects &quot;binomial&quot; as one specific column name in the Polygons data frame
values(shrew)</code></pre>
<pre><code>##   id_no      binomial presence origin seasonal compiler yrcompiled
## 1 29660 Sorex alpinus        1      1        1     IUCN       2008
##                                                citation source dist_comm island
## 1 IUCN (International Union for Conservation of Nature)   &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;
##   subspecies subpop            legend tax_comm  kingdom   phylum    class
## 1       &lt;NA&gt;   &lt;NA&gt; Extant (resident)     &lt;NA&gt; ANIMALIA CHORDATA MAMMALIA
##         order_    family genus category marine terrestial freshwater SHAPE_Leng
## 1 EULIPOTYPHLA SORICIDAE Sorex       NT  False       True      False   88.67521
##   SHAPE_Area
## 1   57.94637</code></pre>
<pre class="r"><code>values(monkey_puzzle_terra)</code></pre>
<pre><code>##              species  gid
## 1 Araucaria_araucana 6382</code></pre>
<pre class="r"><code># For data obtained from the &quot;BIEN&quot; package, we need to convert the sf object into a SpatialPolygonsDataFrame as lets.presab requires this type of object
monkey_puzzle_raster &lt;- as(monkey_puzzle_sf, &quot;Spatial&quot;)

# As the BIEN data do not contain &quot;binomial&quot; as column name, we have to add it:
colnames(monkey_puzzle_raster@data) &lt;- &quot;binomial&quot;

# We set the resolution to 1 degree (the default) and restrict the spatial extent to South America
r_monkey_puzzle &lt;- lets.presab(monkey_puzzle_raster, resol=1, xmn = -100, xmx = -10, ymn = -57, ymx = 15)

# This time, we receive a raster object (instead of SpatRaster)
r_monkey_puzzle</code></pre>
<pre><code>## 
## Class: PresenceAbsence 
## Number of species: 1 
## Number of cells: 243
## Resolution: 1, 1 (x, y)</code></pre>
<pre class="r"><code># Map the range and range centroid
maps::map(&#39;world&#39;,xlim = c(-100,-10),ylim = c(-60,15))
plot(monkey_puzzle_SAm,col=&#39;blue&#39;,add=T)
plot(r_monkey_puzzle, add=T, alpha=0.6, legend=F)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<div class="alert alert-success" role="alert">
<p><strong>Test it yourself</strong></p>
<ul>
<li>Rasterise the range map of the mammal species you selected above.
Try both approaches using the <code>terra</code> package and the
<code>letsR</code> package.</li>
</ul>
</div>
</div>
<div id="bulk-rasterising-multiple-species-range-maps-with-letsr"
class="section level3" number="2.2.3">
<h3><span class="header-section-number">2.2.3</span> Bulk-rasterising
multiple species range maps with <code>letsR</code></h3>
<p>The <code>letsR</code> package also allows to bulk-download multiple
species and rasterise them to form a richness map. As first example, we
look at the <em>Pinus</em> genus in the Americas</p>
<pre class="r"><code># Extract the available Pinus species names
(pinus_names &lt;- BIEN_ranges_genus(&quot;Pinus&quot;,match_names_only = T)[,1])</code></pre>
<pre><code>##  [1] &quot;Pinus_albicaulis&quot;     &quot;Pinus_aristata&quot;       &quot;Pinus_arizonica&quot;     
##  [4] &quot;Pinus_armandii&quot;       &quot;Pinus_attenuata&quot;      &quot;Pinus_ayacahuite&quot;    
##  [7] &quot;Pinus_balfouriana&quot;    &quot;Pinus_banksiana&quot;      &quot;Pinus_brutia&quot;        
## [10] &quot;Pinus_bungeana&quot;       &quot;Pinus_canariensis&quot;    &quot;Pinus_caribaea&quot;      
## [13] &quot;Pinus_cembroides&quot;     &quot;Pinus_clausa&quot;         &quot;Pinus_contorta&quot;      
## [16] &quot;Pinus_coulteri&quot;       &quot;Pinus_culminicola&quot;    &quot;Pinus_densiflora&quot;    
## [19] &quot;Pinus_devoniana&quot;      &quot;Pinus_douglasiana&quot;    &quot;Pinus_durangensis&quot;   
## [22] &quot;Pinus_echinata&quot;       &quot;Pinus_edulis&quot;         &quot;Pinus_elliottii&quot;     
## [25] &quot;Pinus_engelmannii&quot;    &quot;Pinus_flexilis&quot;       &quot;Pinus_georginae&quot;     
## [28] &quot;Pinus_glabra&quot;         &quot;Pinus_greggii&quot;        &quot;Pinus_halepensis&quot;    
## [31] &quot;Pinus_hartwegii&quot;      &quot;Pinus_herrerae&quot;       &quot;Pinus_jeffreyi&quot;      
## [34] &quot;Pinus_koraiensis&quot;     &quot;Pinus_lambertiana&quot;    &quot;Pinus_lawsonii&quot;      
## [37] &quot;Pinus_leiophylla&quot;     &quot;Pinus_longaeva&quot;       &quot;Pinus_luchuensis&quot;    
## [40] &quot;Pinus_lumholtzii&quot;     &quot;Pinus_luzmariae&quot;      &quot;Pinus_maximartinezii&quot;
## [43] &quot;Pinus_maximinoi&quot;      &quot;Pinus_monophylla&quot;     &quot;Pinus_montezumae&quot;    
## [46] &quot;Pinus_monticola&quot;      &quot;Pinus_mugo&quot;           &quot;Pinus_muricata&quot;      
## [49] &quot;Pinus_nelsonii&quot;       &quot;Pinus_nigra&quot;          &quot;Pinus_oocarpa&quot;       
## [52] &quot;Pinus_palustris&quot;      &quot;Pinus_parviflora&quot;     &quot;Pinus_patula&quot;        
## [55] &quot;Pinus_peuce&quot;          &quot;Pinus_pinaster&quot;       &quot;Pinus_pinceana&quot;      
## [58] &quot;Pinus_pinea&quot;          &quot;Pinus_ponderosa&quot;      &quot;Pinus_praetermissa&quot;  
## [61] &quot;Pinus_pringlei&quot;       &quot;Pinus_pseudostrobus&quot;  &quot;Pinus_pumila&quot;        
## [64] &quot;Pinus_pungens&quot;        &quot;Pinus_quadrifolia&quot;    &quot;Pinus_radiata&quot;       
## [67] &quot;Pinus_remota&quot;         &quot;Pinus_resinosa&quot;       &quot;Pinus_rigida&quot;        
## [70] &quot;Pinus_roxburghii&quot;     &quot;Pinus_rzedowskii&quot;     &quot;Pinus_sabiniana&quot;     
## [73] &quot;Pinus_serotina&quot;       &quot;Pinus_strobiformis&quot;   &quot;Pinus_strobus&quot;       
## [76] &quot;Pinus_sylvestris&quot;     &quot;Pinus_tabuliformis&quot;   &quot;Pinus_taeda&quot;         
## [79] &quot;Pinus_tecunumanii&quot;    &quot;Pinus_teocote&quot;        &quot;Pinus_thunbergii&quot;    
## [82] &quot;Pinus_torreyana&quot;      &quot;Pinus_virginiana&quot;     &quot;Pinus_wallichiana&quot;</code></pre>
<pre class="r"><code># Download the range maps for all Pinus species - stored in sf object
(pinus &lt;- BIEN_ranges_load_species(pinus_names))</code></pre>
<pre><code>## Simple feature collection with 84 features and 2 fields
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -179.1256 ymin: -55.95014 xmax: 179.8453 ymax: 63.61258
## Geodetic CRS:  WGS 84
## First 10 features:
##              species   gid                       geometry
## 1   Pinus_albicaulis 69305 MULTIPOLYGON (((-115.5687 3...
## 2     Pinus_aristata 69306 MULTIPOLYGON (((-108.3352 2...
## 3    Pinus_arizonica 69307 MULTIPOLYGON (((-101.8346 1...
## 4     Pinus_armandii 69308 MULTIPOLYGON (((-73.89339 4...
## 5    Pinus_attenuata 69309 MULTIPOLYGON (((-108.6828 2...
## 6   Pinus_ayacahuite 69310 MULTIPOLYGON (((-111.1281 3...
## 7  Pinus_balfouriana 69311 MULTIPOLYGON (((-113.7281 2...
## 8    Pinus_banksiana 69312 MULTIPOLYGON (((-66.87649 -...
## 9       Pinus_brutia 69313 MULTIPOLYGON (((-117.0827 3...
## 10    Pinus_bungeana 69319 MULTIPOLYGON (((-123.3073 4...</code></pre>
<pre class="r"><code># Convert the sf object into a SpatialPolygonsDataFrame to be able to use the lets.presab function
pinus_raster &lt;- as(pinus, &quot;Spatial&quot;)

# As the BIEN data do not contain &quot;binomial&quot; as column name, we have to re-format the column names again before rasterising
colnames(pinus_raster@data) &lt;- &quot;binomial&quot;
r_pinus &lt;- lets.presab(pinus_raster, resol=1, xmn = -170, xmx = -10, ymn = -57, ymx = 60)

# Plot species richness
plot(r_pinus)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>In the <code>lets.presab()</code> function, we can also specify which
PRESENCE, ORIGIN and SEASONAL information should be used and which not
(see IUCN metadata on moodle). Let’s look at Neotropical fruit bats
(<strong>Artibeus</strong>) as example. Here, we set
<code>presence=1</code> meaning that we only consider extant species,
<code>origin=1</code> meaning only native species, and
<code>seasonal=1</code> meaning only resident species.</p>
<p>As the <code>letsR</code> package does not work with
<code>SpatRaster</code> objects yet, we have to use the older
<code>raster</code> package to prepare the range map data.</p>
<!-- ```{r echo=F} -->
<!-- (mammals_raster <- raster::shapefile(paste0(path,'MAMMTERR.shp'))) -->
<!-- ``` -->
<!-- ```{r echo=F, eval=F} -->
<!-- library(raster) -->
<!-- (mammals_raster <- as(mammals,"Spatial")) -->
<!-- # first subset SpatVector and then convert to SpatialPolygonsDataFrame -->
<!-- artibeus_spp <- as(mammals[grep('Artibeus',mammals$BINOMIAL),], "Spatial") -->
<!-- ``` -->
<!-- ```{r eval=F, echo=F} -->
<!-- # prepare raster version of the mammals shapefile -->
<!-- (mammals_raster <- raster::shapefile('data/MAMMTERR.shp')) -->
<!-- ``` -->
<pre class="r"><code>library(raster)

# Subset the SpatVector
artibeus_ranges &lt;- mammals[grep(&#39;Artibeus&#39;,mammals$BINOMIAL),]

# Convert to SpatialPolygonsDataFrame from raster package
artibeus_ranges &lt;- as(artibeus_ranges, &quot;Spatial&quot;)

# Rasterize the ranges using the letsR package
r_artibeus &lt;- lets.presab(artibeus_ranges, resol=2, 
                              presence = 1, origin = 1, seasonal = 1)

# Map the species richness
plot(r_artibeus)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<pre class="r"><code># Map single species - here, just the first two
par(mfrow=c(1,2))
plot(r_artibeus, name = &quot;Artibeus amplus&quot;)
plot(r_artibeus, name = &quot;Artibeus anderseni&quot;)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-18-2.png" width="672" /></p>
<pre class="r"><code># Look at structure of the object and at the presence-absence matrix
str(r_artibeus, 1)</code></pre>
<pre><code>## List of 3
##  $ Presence_and_Absence_Matrix: num [1:339, 1:22] -109 -109 -107 -105 -99 -83 -81 -105 -103 -99 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##  $ Richness_Raster            :Formal class &#39;RasterLayer&#39; [package &quot;raster&quot;] with 13 slots
##  $ Species_name               : chr [1:20] &quot;Artibeus amplus&quot; &quot;Artibeus anderseni&quot; &quot;Artibeus aztecus&quot; &quot;Artibeus cinereus&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;PresenceAbsence&quot;</code></pre>
<pre class="r"><code>head(r_artibeus$Presence_and_Absence_Matrix)</code></pre>
<pre><code>##      Longitude(x) Latitude(y) Artibeus amplus Artibeus anderseni
## [1,]         -109          27               0                  0
## [2,]         -109          25               0                  0
## [3,]         -107          25               0                  0
## [4,]         -105          23               0                  0
## [5,]          -99          23               0                  0
## [6,]          -83          23               0                  0
##      Artibeus aztecus Artibeus cinereus Artibeus concolor Artibeus fimbriatus
## [1,]                0                 0                 0                   0
## [2,]                0                 0                 0                   0
## [3,]                0                 0                 0                   0
## [4,]                1                 0                 0                   0
## [5,]                1                 0                 0                   0
## [6,]                0                 0                 0                   0
##      Artibeus fraterculus Artibeus glaucus Artibeus gnomus Artibeus hirsutus
## [1,]                    0                0               0                 1
## [2,]                    0                0               0                 1
## [3,]                    0                0               0                 1
## [4,]                    0                0               0                 0
## [5,]                    0                0               0                 0
## [6,]                    0                0               0                 0
##      Artibeus incomitatus Artibeus inopinatus Artibeus jamaicensis
## [1,]                    0                   0                    0
## [2,]                    0                   0                    0
## [3,]                    0                   0                    0
## [4,]                    0                   0                    1
## [5,]                    0                   0                    1
## [6,]                    0                   0                    1
##      Artibeus lituratus Artibeus obscurus Artibeus phaeotis
## [1,]                  0                 0                 0
## [2,]                  0                 0                 0
## [3,]                  1                 0                 0
## [4,]                  1                 0                 1
## [5,]                  0                 0                 0
## [6,]                  0                 0                 0
##      Artibeus planirostris Artibeus rosenbergii Artibeus toltecus
## [1,]                     0                    0                 0
## [2,]                     0                    0                 0
## [3,]                     0                    0                 1
## [4,]                     0                    0                 1
## [5,]                     0                    0                 1
## [6,]                     0                    0                 0
##      Artibeus watsoni
## [1,]                0
## [2,]                0
## [3,]                0
## [4,]                0
## [5,]                0
## [6,]                0</code></pre>
</div>
<div id="bulk-rasterising-multiple-species-range-maps-with-terra"
class="section level3" number="2.2.4">
<h3><span class="header-section-number">2.2.4</span> Bulk-rasterising
multiple species range maps with <code>terra</code></h3>
<p>Bulk-rasterising with <code>terra</code> is not as straight forward
but requires some legwork.</p>
<pre class="r"><code>library(tidyverse)

# Extract the available Pinus species names
pinus_names &lt;- BIEN_ranges_genus(&quot;Pinus&quot;,match_names_only = T)[,1]

# Download the range maps for all Pinus species
pinus &lt;- BIEN_ranges_load_species(pinus_names)

# For sf objects, we can use tidyverse to subset single features (species)
filter(pinus, species==&quot;Pinus_albicaulis&quot;)</code></pre>
<pre><code>## Simple feature collection with 1 feature and 2 fields
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -126.2099 ymin: 30.86017 xmax: -105.0399 ymax: 51.65776
## Geodetic CRS:  WGS 84
##            species   gid                       geometry
## 1 Pinus_albicaulis 69305 MULTIPOLYGON (((-115.5687 3...</code></pre>
<pre class="r"><code># Use &quot;sapply&quot; function to circle through all species and rasterise to 1 degree spatial resolution:
r_1deg &lt;- terra::rast(res=1)
r_pinus_terra &lt;- sapply(unique(pinus$species),FUN=function(sp) {
  terra::rasterize(filter(pinus, species==sp), r_1deg, touches=T) 
  })

# Make a SpatRaster stack with all species occurrences
r_pinus_terra &lt;- terra::rast(r_pinus_terra)

# Calculate species richness
r_pinus_richness &lt;- sum(r_pinus_terra,na.rm=T)

# Plot species richness
maps::map(&#39;world&#39;)
plot(r_pinus_richness, add=T, alpha=0.8)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<pre class="r"><code># We can collate a data.frame with all species occurrences similar to the lets.presab output
r_pinus_df &lt;- as.data.frame(r_pinus_terra, xy=T)</code></pre>
<p>Let’s also look at the mammal example. Remember that we wanted to
subset the data according to the IUCN information PRESENCE, ORIGIN and
SEASONAL.</p>
<pre class="r"><code># Subset the SpatVector
artibeus_ranges &lt;- mammals[grep(&#39;Artibeus&#39;,mammals$BINOMIAL),]
artibeus_ranges &lt;- subset(artibeus_ranges, artibeus_ranges$PRESENCE==1 &amp; artibeus_ranges$ORIGIN==1 &amp; artibeus_ranges$SEASONAL==1)

# Artibeus species:
unique(artibeus_ranges$BINOMIAL)</code></pre>
<pre><code>##  [1] &quot;Artibeus amplus&quot;       &quot;Artibeus anderseni&quot;    &quot;Artibeus aztecus&quot;     
##  [4] &quot;Artibeus cinereus&quot;     &quot;Artibeus concolor&quot;     &quot;Artibeus fimbriatus&quot;  
##  [7] &quot;Artibeus fraterculus&quot;  &quot;Artibeus glaucus&quot;      &quot;Artibeus gnomus&quot;      
## [10] &quot;Artibeus hirsutus&quot;     &quot;Artibeus incomitatus&quot;  &quot;Artibeus inopinatus&quot;  
## [13] &quot;Artibeus jamaicensis&quot;  &quot;Artibeus lituratus&quot;    &quot;Artibeus obscurus&quot;    
## [16] &quot;Artibeus phaeotis&quot;     &quot;Artibeus planirostris&quot; &quot;Artibeus rosenbergii&quot; 
## [19] &quot;Artibeus toltecus&quot;     &quot;Artibeus watsoni&quot;</code></pre>
<pre class="r"><code># As we saw above, for SpatVector objects, we can use &quot;subset&quot; to select specific rows
subset(artibeus_ranges, artibeus_ranges$BINOMIAL==&quot;Artibeus amplus&quot;)</code></pre>
<pre><code>##  class       : SpatVector 
##  geometry    : polygons 
##  dimensions  : 1, 14  (geometries, attributes)
##  extent      : -77.9345, -55.3096, 0.9759356, 12.46334  (xmin, xmax, ymin, ymax)
##  coord. ref. : lon/lat WGS 84 (EPSG:4326) 
##  names       :        BINOMIAL PRESENCE ORIGIN SEASONAL COMPILER  YEAR
##  type        :           &lt;chr&gt;    &lt;int&gt;  &lt;int&gt;    &lt;int&gt;    &lt;chr&gt; &lt;int&gt;
##  values      : Artibeus amplus        1      1        1     IUCN  2008
##         CITATION SOURCE DIST_COMM ISLAND (and 4 more)
##            &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;             
##  IUCN (Internat~     NA        NA     NA</code></pre>
<pre class="r"><code># Use &quot;sapply&quot; function to circle through all species and rasterise to 1 degree spatial resolution:
r_1deg &lt;- terra::rast(res=1)
r_artibeus_terra &lt;- sapply(unique(artibeus_ranges$BINOMIAL),FUN=function(sp) {
  terra::rasterize(subset(artibeus_ranges, artibeus_ranges$BINOMIAL==sp), r_1deg, touches=T) 
  })

# Make a SpatRaster stack with all species occurrences
r_artibeus_terra &lt;- terra::rast(r_artibeus_terra)

# Calculate species richness
r_artibeus_richness &lt;- sum(r_artibeus_terra,na.rm=T)

# Map the species richness
maps::map(&#39;world&#39;)
plot(r_artibeus_richness, add=T, alpha=0.8)</code></pre>
<p><img src="a4_RangeMaps_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<div class="alert alert-info">
<p><em><strong>Exercise:</strong></em></p>
<ul>
<li>Select another genus of plants or mammals, and follow the workflow
to rasterise range maps (either using <code>terra</code> or
<code>letsR</code>) and map species richness of the species within that
genus.</li>
<li>Plot the latitudinal species richness gradient for this genus.
Interpret in light of previously discussed patterns.</li>
</ul>
</div>
</div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Hurlbert2007" class="csl-entry">
Hurlbert, A. H., and W. Jetz. 2007. <span>“Species Richness, Hotspots,
and the Scale Dependence of Range Maps in Ecology and
Conservation.”</span> <em>PNAS</em> 104: 13384–89.
</div>
<div id="ref-Jetz2012a" class="csl-entry">
Jetz, W., J. M. McPherson, and R. P. Guralnick. 2012. <span>“Integrating
Biodiversity Distribution Knowledge: Toward a Global Map of
Life.”</span> <em>Trends in Ecology &amp; Evolution</em> 27: 151–59.
</div>
<div id="ref-Krosby2015" class="csl-entry">
Krosby, M., C. B. Wilsey, J. L. McGuire, J. M. Duggan, T. M. Nogeire, J.
A. Heinrichs, J. J. Tewksbury, and J. J. Lawler. 2015.
<span>“Climate-Induced Range Overlap Among Closely Related
Species.”</span> <em>Nature Climate Change</em> 5: 883–86.
</div>
<div id="ref-W_est_2020" class="csl-entry">
Wüest, R. O., N. E. Zimmermann, D. Zurell, J. M. Alexander, S. A. Fritz,
C. Hof, H. Kreft, et al. 2020. <span>“Macroecology in the Age of Big
Data <span></span> Where to Go from Here?”</span> <em>Journal of
Biogeography</em> 47 (1): 1–12. <a
href="https://doi.org/10.1111/jbi.13633">https://doi.org/10.1111/jbi.13633</a>.
</div>
<div id="ref-Zurell2018c" class="csl-entry">
Zurell, D., C. H. Graham, L. Gallien, W. Thuiller, and N. E. Zimmermann.
2018. <span>“Long-Distance Migratory Birds Threatened by Multiple
Independent Risks from Global Change.”</span> <em>Nature Climate
Change</em> 8 (October): 992–96. <a
href="https://doi.org/10.1038/s41558-018-0312-9">https://doi.org/10.1038/s41558-018-0312-9</a>.
</div>
</div>
</div>

<!DOCTYPE html>
<html>

<br>
<hr />
<div id="footer">
<p>Damaris Zurell 2022 <a href="http://creativecommons.org/licenses/by/4.0/" >(CC BY 4.0)</a>.  </p>
</div>

</html>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
