<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>SDM assessment and prediction</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Macroecology and global change</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pracs: macroecological analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="a0_setup.html">0. Getting started</a>
    </li>
    <li>
      <a href="a1_SpatialData.html">1. Spatial data in R</a>
    </li>
    <li>
      <a href="a2_RichnessGradients.html">2. Species richness gradients</a>
    </li>
    <li>
      <a href="a3_RichnessRegression.html">3. Species richness regression</a>
    </li>
    <li>
      <a href="a4_RangeMaps.html">4. Species range maps</a>
    </li>
    <li>
      <a href="a5_SpeciesThreats.html">5. Species threats</a>
    </li>
    <li>
      <a href="a6_BiodivChanges.html">6. Analysing biodiversity changes</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pracs: species distribution modelling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="b1_SpeciesData.html">1. Species data</a>
    </li>
    <li>
      <a href="b2_EnvData.html">2. Environmental data</a>
    </li>
    <li>
      <a href="b3_SDM_intro.html">3. SDMs: simple model fitting</a>
    </li>
    <li>
      <a href="b4_SDM_eval.html">4. SDMs: assessment and prediction</a>
    </li>
    <li>
      <a href="b5_pseudoabsence.html">5. Pseudo-absence and background data</a>
    </li>
    <li>
      <a href="b6_SDM_algorithms.html">6. SDMs: algorithms</a>
    </li>
    <li>
      <a href="b7_SDM_ensembles.html">7. SDMs: ensembles</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.uni-potsdam.de/en/ibb-macroecology/index">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/ZurellLab">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">SDM assessment and prediction</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div class="alert alert-info">
<p><strong>RStudio project</strong></p>
<p>Open the RStudio project that we created in the first session. I
recommend to use this RStudio project for the entire course and within
the RStudio project create separate R scripts for each session.</p>
<ul>
<li>Create a new empty R script by going to the tab “File”, select “New
File” and then “R script”</li>
<li>In the new R script, type
<code># Session b4: SDM assessment and prediction</code> and save the
file in your folder “scripts” within your project folder, e.g. as
“b4_SDM_eval.R”</li>
</ul>
</div>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>We have already fitted simple species distribution models (SDMs) in
the previous tutorial. Remember the five general model building steps:
(i) conceptualisation, (ii) data preparation, (iii) model fitting, (iv)
model assessment, and (v) prediction <span class="citation">(Zurell et
al. 2020)</span>. In this practical, we will concentrate on steps
(iv)-(v). We will learn how to validate SDMs, visualise the fitted
species-environment relationships, and make predictions. For getting a
deeper understanding of these single steps, I highly recommend studying
more advanced reviews <span class="citation">(Guisan and Zimmermann
2000; Guisan and Thuiller 2005; Elith and Leathwick 2009)</span> and
textbooks on SDMs <span class="citation">(Franklin 2010; Guisan,
Thuiller, and Zimmermann 2017)</span>.</p>
<div id="recap-of-last-session-data-and-models" class="section level2"
number="1.1">
<h2><span class="header-section-number">1.1</span> Recap of last
session: data and models</h2>
<p>We will continue to work on the Ring Ouzel example of the previous
session, using data from the British breeding and wintering birds
citizen science atlas <span class="citation">(Gillings et al.
2019)</span>. Let’s quickly repeat the main steps of data preparation
and model fitting:</p>
<pre class="r"><code>library(terra)
bg &lt;- terra::rast(&#39;data/Prac3_UK_mask.grd&#39;)
sp_dat &lt;- read.table(&#39;data/Prac3_RingOuzel.txt&#39;,header=T)</code></pre>
<pre class="r"><code>m1 &lt;- step(glm( Turdus_torquatus ~ bio1 + I(bio1^2) + bio8 + I(bio8^2), family=&#39;binomial&#39;, data=sp_dat))</code></pre>
<pre><code>## Start:  AIC=132.24
## Turdus_torquatus ~ bio1 + I(bio1^2) + bio8 + I(bio8^2)
## 
##             Df Deviance    AIC
## - bio1       1   122.25 130.25
## - I(bio1^2)  1   122.30 130.30
## - bio8       1   123.64 131.64
## &lt;none&gt;           122.24 132.24
## - I(bio8^2)  1   124.36 132.36
## 
## Step:  AIC=130.25
## Turdus_torquatus ~ I(bio1^2) + bio8 + I(bio8^2)
## 
##             Df Deviance    AIC
## &lt;none&gt;           122.25 130.25
## - bio8       1   124.86 130.86
## - I(bio8^2)  1   126.00 132.00
## - I(bio1^2)  1   140.38 146.38</code></pre>
<pre class="r"><code>summary(m1)</code></pre>
<pre><code>## 
## Call:
## glm(formula = Turdus_torquatus ~ I(bio1^2) + bio8 + I(bio8^2), 
##     family = &quot;binomial&quot;, data = sp_dat)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -1.42982  -0.33815  -0.11284  -0.02967   3.00631  
## 
## Coefficients:
##             Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)  1.76080    1.66925   1.055  0.29150    
## I(bio1^2)   -0.10745    0.02917  -3.684  0.00023 ***
## bio8         1.61773    1.04632   1.546  0.12208    
## I(bio8^2)   -0.20389    0.11734  -1.738  0.08229 .  
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 195.68  on 302  degrees of freedom
## Residual deviance: 122.25  on 299  degrees of freedom
## AIC: 130.25
## 
## Number of Fisher Scoring iterations: 8</code></pre>
</div>
</div>
<div id="model-assessment" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Model assessment</h1>
<p>Before we can use our model for making predictions in space and time,
we need to assess model behaviour and predictive performance.</p>
<div id="visualising-response-curves" class="section level2"
number="2.1">
<h2><span class="header-section-number">2.1</span> Visualising response
curves</h2>
<p>When only looking at parameter estimates, it is sometimes difficult
to envision how exactly the fitted response (the “niche” function) looks
like. Also, for more complicated machine learning algorithms (that we
will get to know later), there are no parameter estimates to look at, so
we need different means to judge the plausibility of the fitted
response.</p>
<p>Response curves and response surfaces visualize the (mean) values
that a model would predict for an environmental situation, meaning for
specific values of the predictor variables. In the Ring Ouzel example,
we have only two predictors, so we can simply construct a 3D surface
that shows the predicted values along the two environmental
gradients.</p>
<p>We can get the predicted values using the <code>predict()</code>
function.</p>
<pre class="r"><code># If we do not provide &quot;newdata&quot;, then predict() should simply return the fitted values: 
head(predict(m1, type=&#39;response&#39;))</code></pre>
<pre><code>##            1            2            3            4            5            6 
## 8.631855e-02 5.658431e-03 2.207752e-02 2.036298e-02 2.878656e-05 4.897743e-02</code></pre>
<pre class="r"><code>head(m1$fitted)</code></pre>
<pre><code>##            1            2            3            4            5            6 
## 8.631855e-02 5.658431e-03 2.207752e-02 2.036298e-02 2.878656e-05 4.897743e-02</code></pre>
<p>If we want to predict model response along the two environmental
gradients, we first need to define a grid that contains all combinations
of the two variables. For this, we use a combination of the
<code>expand.grid()</code> and <code>seq()</code> functions.</p>
<pre class="r"><code># Wwe want to make predictions for all combinations of the two predictor variables
# and along their entire environmental gradients:
xyz &lt;- expand.grid(
  # We produce a sequence of environmental values within the predictor ranges:
    bio1 = seq(min(sp_dat$bio1),max(sp_dat$bio1),length=50),
    bio8 = seq(min(sp_dat$bio8),max(sp_dat$bio8),length=50)
    )

# Now we can make predictions to this new data frame
xyz$z &lt;- predict(m1, newdata=xyz, type=&#39;response&#39;)
summary(xyz)</code></pre>
<pre><code>##       bio1             bio8              z            
##  Min.   : 5.274   Min.   : 1.090   Min.   :0.0000000  
##  1st Qu.: 6.637   1st Qu.: 4.107   1st Qu.:0.0000174  
##  Median : 8.058   Median : 7.251   Median :0.0023505  
##  Mean   : 8.058   Mean   : 7.251   Mean   :0.1112626  
##  3rd Qu.: 9.478   3rd Qu.:10.395   3rd Qu.:0.0881662  
##  Max.   :10.841   Max.   :13.413   Max.   :0.8785273</code></pre>
<pre class="r"><code># As result, we have a 3D data structure and want to visualise this.
# Here, I first set a color palette
library(RColorBrewer)
cls &lt;- colorRampPalette(rev(brewer.pal(11, &#39;RdYlBu&#39;)))(100)

# Finally, we plot the response surface using the wireframe function from the lattice package
library(lattice)
wireframe(z ~ bio1 + bio8, data = xyz, zlab = list(&quot;Occurrence prob.&quot;, rot=90), 
          drape = TRUE, col.regions = cls, scales = list(arrows = FALSE), zlim = c(0, 1))</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code># We can also rotate the axes to better see the surface
wireframe(z ~ bio1 + bio8, data = xyz, zlab = list(&quot;Occurrence prob.&quot;, rot=90), 
          drape = TRUE, col.regions = cls, scales = list(arrows = FALSE), zlim = c(0, 1), 
          screen=list(z = -160, x = -70, y = 3))</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<p>This looks very nice. However, if the model gets more complicated and
contains more variables than just two, how can we then visualise it?</p>
<p>One way is to cut the response surface into slices. Most often, we
simply plot the response along one environmental gradient while keeping
all other gradients constant at their mean. We call this
<strong><em>partial response plots</em></strong>. For simplicity, we can
use the function <code>partial_response()</code> from the
<code>mecofun</code> package for plotting.</p>
<pre class="r"><code>library(mecofun)

# Names of our variables:
my_preds &lt;- c(&#39;bio1&#39;, &#39;bio8&#39;)

# We want two panels next to each other:
par(mfrow=c(1,2))

# Plot the partial responses
partial_response(m1, predictors = sp_dat[,my_preds], ylab=&#39;Occurrence probability&#39;)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>When you compare the response surface and the partial response
curves, you see that the latter only give us an incomplete picture of
what is going on as they are only “slicing” the response surface in the
middle.</p>
<div class="alert alert-info">
<p><em><strong>Exercise:</strong></em></p>
<p>Fit a GLM with only two different (weakly correlated) predictor
variables for the Yellowhammer data (previous prac) and store as object
(e.g. as <code>m_yellowhammer</code>). Plot the response surface, and
the partial response curves.</p>
<ul>
<li>How do you interpret the fitted relationship?</li>
</ul>
</div>
</div>
<div id="assessing-sdm-performance" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Assessing SDM
performance</h2>
<p>In the previous session, we already learned about the measure
“explained deviance” that tells us something about the goodness-of-fit,
meaning how well the model fits the data.</p>
<pre class="r"><code>expl_deviance(obs = sp_dat$Turdus_torquatus,
              pred = m1$fitted)</code></pre>
<pre><code>## [1] 0.3752566</code></pre>
<p>Here, our two-predictor model explains roughly 40% of the variance in
the data. However, often we are not only interested in how well our
model fits the data but how robust the model is against changes in the
input data and, thus, how robust predictions to other places and times
might be. This assessment of model performance is often referred to as
validation or evaluation. Evaluating the model on the calibration or
training data is often referred to as internal validation
(“resubstitution”) <span class="citation">(Araujo et al. 2005)</span>.
This generally gives a too optimistic picture of model performance. It
is better to evaluate the model using data that have not been used for
model fitting. One way is to split the data and only train the model on
some proportion of the data and validate it on the hold-out data.
Potential procedures are repeated split-samples (e.g. splitting into 70%
training and 30% test data and repeat several times) and <em>k</em>-fold
cross-validation (e.g. 5-fold or 10-fold), where the data are split into
<em>k</em> portions, the <em>k</em>th portion is held out for validation
and the procedure is repeated <em>k</em> times. If these folds are
stratified in geographic or environmental space, we talk of spatial
block cross-validation and environmental block cross-validation <span
class="citation">(Roberts et al. 2017)</span>.</p>
<p>Here, we will use the function <code>crossvalSDM()</code> from the
<code>mecofun</code> package to split our data into 5 folds,
re-calibrate the model using only 4/5 of the original data and predict
the model to the hold-out 1/5 of the data.</p>
<pre class="r"><code>preds_cv &lt;- crossvalSDM(m1, traindat = sp_dat, colname_species = &#39;Turdus_torquatus&#39;, colname_pred = my_preds)</code></pre>
<p>We receive a numeric vector of cross-validated prediction. Out of
curiosity, let us compare the fitted values on the training data and the
predictions on the cross-validation data. You should see that the
predictions are similar (this is good, otherwise our model would be very
sensitive to changes in input data) but not identical - thus, we have
basically added some noise.</p>
<pre class="r"><code>plot(m1$fitted.values, preds_cv, xlab=&#39;Fitted values&#39;, ylab=&#39;Predicted values from CV&#39;)
abline(0,1,col=&#39;red&#39;,lwd=2)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>We will use these cross-validated predictions to assess model
performance.</p>
<div id="threshold-dependent-performance-measures"
class="section level3" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Threshold-dependent
performance measures</h3>
<p>Now, we want to know how well our model predicts the observations.
Different measures are available for quantifying this. A lot of these
measures are threshold-dependent. You have probably realised that our
model predicts a continuous response, the probability of occurrence,
while our observations are binary. Many performance measures rely on
comparisons like “How many presence observations does the model
correctly predict as presence”. In order to answer this we first need to
convert the continuous probabilities into binary predictions. Different
thresholds are introduced in <span class="citation">Liu et al.
(2005)</span>. Most of these are implemented in the
<code>PresenceAbsence</code> package in the
<code>optimal.thresholds</code> function.</p>
<pre class="r"><code>library(PresenceAbsence)

# We first prepare our data:
# Prepare cross-validated predictions:
thresh_dat &lt;- data.frame(
  ID = seq_len(nrow(sp_dat)), 
    obs = sp_dat$Turdus_torquatus,
    pred = preds_cv)
        
# Then, we find the optimal thresholds:     
(thresh_cv &lt;- PresenceAbsence::optimal.thresholds(DATA= thresh_dat))</code></pre>
<pre><code>##          Method      pred
## 1       Default 0.5000000
## 2     Sens=Spec 0.1550000
## 3  MaxSens+Spec 0.1550000
## 4      MaxKappa 0.2200000
## 5        MaxPCC 0.5300000
## 6  PredPrev=Obs 0.3700000
## 7       ObsPrev 0.0990099
## 8      MeanProb 0.1001586
## 9    MinROCdist 0.1550000
## 10      ReqSens 0.1600000
## 11      ReqSpec 0.1500000
## 12         Cost 0.5300000</code></pre>
<p>We can compare observed vs. predicted presences and absences based on
these tresholds. For this, we take our predictions from the
cross-validation. The comparison is easiest illustrated in a confusion
matrix, for example using the function <code>cmx</code> in the
<code>PresenceAbsence</code> package.</p>
<p>Have a look at <span class="citation">Liu et al. (2005)</span> to see
which thresholds they recommend. Here, we will use the threshold that
maximises the sum of sensitivity and specificity (the third row in the
thresholds data frame):</p>
<pre class="r"><code>(cmx_maxSSS &lt;- PresenceAbsence::cmx(DATA= thresh_dat, threshold=thresh_cv[3,2]))</code></pre>
<pre><code>##          observed
## predicted   1   0
##         1  26  40
##         0   4 233</code></pre>
<p>From such a confusion matrix, we can calculate different evaluation
criteria. For example,<br />
- the proportion of correctly classified test observations
<code>pcc</code><br />
- the proportion of correctly classified presences, also called
sensitivity or true positive rate<br />
- the proportion of correctly classified absences, also called
specificity or true negative rate</p>
<pre class="r"><code># Proportion of correctly classified observations
PresenceAbsence::pcc(cmx_maxSSS, st.dev=F)</code></pre>
<pre><code>## [1] 0.8547855</code></pre>
<pre class="r"><code># Sensitivity = true positive rate
PresenceAbsence::sensitivity(cmx_maxSSS, st.dev=F)</code></pre>
<pre><code>## [1] 0.8666667</code></pre>
<pre class="r"><code># Specificity = true negative rate
PresenceAbsence::specificity(cmx_maxSSS, st.dev=F)</code></pre>
<pre><code>## [1] 0.8534799</code></pre>
<p>Other measures are <em>Kappa</em> and <em>TSS</em> (the true skill
statistic). <span class="citation">Allouche, Tsoar, and Kadmon
(2006)</span> explain how to calculate these.</p>
<pre class="r"><code># Kappa
PresenceAbsence::Kappa(cmx_maxSSS, st.dev=F)</code></pre>
<pre><code>## [1] 0.4694365</code></pre>
<pre class="r"><code># True skill statistic
TSS(cmx_maxSSS) </code></pre>
<pre><code>## [1] 0.7201465</code></pre>
<p>According to <span class="citation">Araujo et al. (2005)</span>,
<em>Kappa</em>&gt;0.4 indicate good predictions. For TSS, we often
assume TSS&gt;0.5 to indicate good predictions.</p>
</div>
<div id="threshold-independent-performance-measures"
class="section level3" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span>
Threshold-independent performance measures</h3>
<p>The most common evaluation statistic that avoids thresholding the
data is AUC - the area under the receiver-operating characteristic (ROC)
curve. ROC curves are generated by calculating sensitivity (true
positive rate) and specificity (true negative rate) for many thresholds
along the entire range of predicted probabilities. Then, (1-specificity)
is plotted on the x-axis against sensitivity on the y axis. The area
under this curve is called the AUC. The further the generated curve
deviates from the 1:1 line towards the upper-left corner, the better the
model predicts presence/absence of a species. If we would take a random
presence and a random absence from our observations and make
predictions, than AUC can be interpeted as the chance of assigning a
higher predicted occurrence probability to the presence compared to the
absence point. Typically, we regard AUC&gt;0.7 as indicating fair
predictions <span class="citation">(Araujo et al. 2005)</span>.</p>
<pre class="r"><code>library(AUC)

# Let&#39;s have a look a the ROC curve:
roc_cv &lt;- roc(preds_cv, as.factor(sp_dat$Turdus_torquatus))
plot(roc_cv, col = &quot;grey70&quot;, lwd = 2)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<pre class="r"><code># Compute the AUC:
AUC::auc(roc_cv)</code></pre>
<pre><code>## [1] 0.9042735</code></pre>
<p>It seems our model is performing pretty well on hold-out data. We can
thus attempt making predictions in space and time.</p>
<p>Please be aware that many packages contain functions for evaluating
SDMs. As always you have to find your own way. Here, I provide merely
examples. To ease further performance assessments during this course,
the <code>mecofun</code> package contains a function
<code>evalSDM()</code> that computes the here-mentioned performance
measures. By default, this function uses the <em>MaxSens+Spec</em>
threshold.</p>
<pre class="r"><code>evalSDM(sp_dat$Turdus_torquatus, preds_cv)</code></pre>
<pre><code>##         AUC       TSS     Kappa      Sens      Spec       PCC        D2 thresh
## 1 0.9042735 0.7201465 0.4694365 0.8666667 0.8534799 0.8547855 0.3507918  0.155</code></pre>
<div class="alert alert-info">
<p><em><strong>Exercise:</strong></em></p>
<p>Use the <code>m_yellowhammer</code> model from the previous exercise
and assess the TSS, sensitivity, specificity, and AUC.</p>
<ul>
<li>The model of which species (<code>m1</code> or
<code>m_yellowhammer</code>) has a higher predictive performance?</li>
</ul>
</div>
</div>
</div>
</div>
<div id="spatio-temporal-predictions" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Spatio-temporal
predictions</h1>
<p>We have already learned how to make predictions using the function
<code>predict()</code> and also using the argument <code>newdata</code>.
All we need for transferring our model to other places and times are the
respective environmental variables.</p>
<div id="prepare-the-environmental-layers" class="section level2"
number="3.1">
<h2><span class="header-section-number">3.1</span> Prepare the
environmental layers</h2>
<p>As you have learned in the previous sessions, we can download current
and future climate layers from databases such as <a
href="http://www.worldclim.org/">WorldClim</a> and <a
href="http://chelsa-climate.org/">CHELSA</a>. When making predictions,
we have to take care that the target environmental layers match the
scale of analysis. Our species data were available at a 10 km spatial
resolution based on the British National Grid. We aim to predict our
model to climate layers based on the worldclim data base, which offers
the data at different resolutions in geographic (lon/lat) projection.
The appropriate worldclim spatial resolution would be the 5 min
resolution. As a necessary step, we need to reproject these climate
layers to the British National Grid. Below, I show you how this could be
done. Alternatively, you can also download the processed climate layers
<a href="data/Prac4_climate.zip">here</a> (or from the moodle) and read
them in using the <code>terra::rast()</code> function.</p>
<p>We use the <code>geodata</code> package for downloading worldclim
data. As we only want to retrieve the data for UK, we can use the
<code>geodata::worldclim_country()</code> function for the current
climate. We download data for the current climate as well as a future
scenario for the year 2050.</p>
<pre class="r"><code>library(geodata)

# Please note that you have to set download=T if you haven&#39;t downloaded the data before:
bio_curr &lt;- geodata::worldclim_country(&#39;GB&#39;, var = &#39;bio&#39;, res = 5, download = F, path = &#39;data&#39;)

# Please note that you have to set download=T if you haven&#39;t downloaded the data before:
bio_fut &lt;- geodata::cmip6_world(model=&#39;ACCESS-ESM1-5&#39;, ssp=&#39;245&#39;, time=&#39;2041-2060&#39;, var=&#39;bioc&#39;, download=F, res=5, path=&#39;data&#39;)</code></pre>
<!-- ```{r} -->
<!-- bio_fut <- geodata::cmip6_world(model='MPI-ESM1-2-LR', ssp='245', time='2041-2060', var='bioc', download=F, res=5, path='data') -->
<!-- ``` -->
<p>We will use the background mask of UK to clip the data. This mask is
in the British National Grid, which is the target coordinate system, and
we thus need to reproject the worldclim layers. To speed things up, we
will first crop the climate layers.</p>
<pre class="r"><code># the approx. spatial extent of UK in lon/lat coordinates
extent_uk &lt;- c(-12, 3, 48, 62)

# Crop and reproject current climate
bio_curr &lt;- terra::crop(bio_curr, extent_uk)
bio_curr &lt;- terra::project(bio_curr, bg)
bio_curr &lt;- terra::mask(bio_curr, bg)

# Crop and reproject future climate
bio_fut &lt;- terra::crop(bio_fut, extent_uk)
bio_fut &lt;- terra::project(bio_fut, bg)
bio_fut &lt;- terra::mask(bio_fut, bg)</code></pre>
<p>The problem now is that the current and the future climate layers
have different variable names than our original species data frame. The
function <code>predict()</code> will only work if all have the same
names as in the model specification. We thus have to change the climate
layer names.</p>
<pre class="r"><code># Change names of climate layers
names(bio_curr) &lt;- names(bio_fut) &lt;- names(sp_dat)[-c(1:3)]</code></pre>
<p>Store the layers for later use:</p>
<pre class="r"><code>writeRaster(bio_curr, &#39;data/Prac4_UK_bio_curr.grd&#39;)
writeRaster(bio_fut, &#39;data/Prac4_UK_bio_fut.grd&#39;)</code></pre>
</div>
<div id="make-predictions-to-the-environmental-layers"
class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Make predictions to
the environmental layers</h2>
<p>It is now straight forward to make continuous predictions to the
current and the future climate:</p>
<pre class="r"><code># Prepare data frames
bio_curr_df &lt;- data.frame(crds(bio_curr),as.points(bio_curr))
bio_fut_df &lt;- data.frame(crds(bio_fut),as.points(bio_fut))

# Make continuous predictions:
bio_curr_df$pred_glm &lt;- predict(m1, newdata= bio_curr_df, type=&quot;response&quot;)
bio_fut_df$pred_glm &lt;- predict(m1, newdata= bio_fut_df, type=&quot;response&quot;)</code></pre>
<p>Of course, we can also plot the predictions:</p>
<pre class="r"><code>par(mfrow=c(1,2))

# Make raster of predictions to current environment:
r_pred_curr &lt;- terra::rast(bio_curr_df[,c(&#39;x&#39;,&#39;y&#39;,&#39;pred_glm&#39;)], type=&#39;xyz&#39;, crs=crs(bg))
plot(r_pred_curr, axes=F, main=&#39;Occ. prob. - today&#39;)

# Make raster stack of predictions to future environment:
r_pred_fut &lt;- terra::rast(bio_fut_df[,c(&#39;x&#39;,&#39;y&#39;,&#39;pred_glm&#39;)], type=&#39;xyz&#39;, crs=crs(bg))
plot(r_pred_fut, axes=F, main=&#39;Occ. prob. - 2050&#39;)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<p>We see that ring ouzel is predicted to decline under the chosen
climate scenario.</p>
<p>Lastly, we can also translate the continuous predictions into binary
predictions and plot the resulting maps.</p>
<pre class="r"><code># Make binary predictions:
bio_curr_df$bin_glm &lt;- ifelse(bio_curr_df$pred_glm &gt;= thresh_cv[3,2], 1, 0)
bio_fut_df$bin_glm &lt;- ifelse(bio_fut_df$pred_glm &gt;= thresh_cv[3,2], 1, 0)

# Make raster stack of predictions to current environment:
r_pred_curr &lt;- terra::rast(bio_curr_df[,c(&#39;x&#39;,&#39;y&#39;,&#39;pred_glm&#39;,&#39;bin_glm&#39;)], type=&#39;xyz&#39;, crs=crs(bg))
plot(r_pred_curr, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<pre class="r"><code># Make raster stack of predictions to future environment:
r_pred_fut &lt;- terra::rast(bio_fut_df[,c(&#39;x&#39;,&#39;y&#39;,&#39;pred_glm&#39;,&#39;bin_glm&#39;)], type=&#39;xyz&#39;, crs=crs(bg))
plot(r_pred_fut, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-22-2.png" width="672" /></p>
<div class="alert alert-info">
<p><em><strong>Exercise:</strong></em></p>
<p>Use the <code>m_yellowhammer</code> model from the previous exercise
and make predictions to current and future climates.</p>
<ul>
<li>For which species does the model predict a larger area under current
climate?</li>
<li>for which species does the model predict a larger difference between
current and future climate?</li>
</ul>
</div>
</div>
<div id="assessing-novel-environments" class="section level2"
number="3.3">
<h2><span class="header-section-number">3.3</span> Assessing novel
environments</h2>
<p>Novel environments are conditions that were not realised in the
sampled data but are realised in the projection data. For example, in
the future it may be warmer than today. If the entire niche of the
species is encompassed by data, then the model does not need to
extrapolate even if the projection data contain some novel environments.
Mostly, novel environments only prove problematic if the niche is
truncated in the sampled data <span class="citation">(Zurell, Elith, and
Schroeder 2012)</span>. Novel environments can be assessed in different
ways. MESS (Multivariate environmental similarity surface) maps are
contained in the <code>predicts</code> package and described in <span
class="citation">Elith, Kearney, and Phillips (2010)</span>. They assess
for each environmental variables separately whether the projection data
contain novel conditions beyond the sampled range.</p>
<pre class="r"><code>library(predicts)

# MESS maps from the predicts package:
r_mess &lt;- mess(bio_fut[[my_preds]], sp_dat[,my_preds])
plot(r_mess, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<pre class="r"><code># Negative values indicate dissimilar=novel environments:
r_mess_mask &lt;- r_mess &lt; 0
plot(r_mess_mask, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-23-2.png" width="672" /></p>
<p>From these maps, we can already see that novel environments should
not be any issue for the ring ouzel as novel environments could mainly
arise in the South while the Ring Ouzel is a northern distributed
species. Nevertheless, we can further explore model projections for
analogous climates versus novel climates:</p>
<pre class="r"><code># Predictions to analogous climates:
r_analog_fut &lt;- terra::extend(r_pred_fut, bg)
values(r_analog_fut)[values(r_mess)&lt;0] &lt;- NA
plot(r_analog_fut, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<pre class="r"><code># Predictions to novel climates:
r_novel_fut &lt;- terra::extend(r_pred_fut, bg)
values(r_novel_fut)[values(r_mess)&gt;=0] &lt;- NA
plot(r_novel_fut, axes=F)</code></pre>
<p><img src="b4_SDM_eval_files/figure-html/unnamed-chunk-24-2.png" width="672" /></p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-allouche2006" class="csl-entry">
Allouche, O., A. Tsoar, and R. Kadmon. 2006. <span>“Assessing the
Accuracy of Species Distribution Models: Prevalence, Kappa and the True
Skill Statistic (TSS).”</span> <em>Journal of Applied Ecology</em> 43:
1223–32.
</div>
<div id="ref-araujo2005" class="csl-entry">
Araujo, M. B., R. G. Pearson, W. Thuiller, and M. Erhard. 2005.
<span>“Validation of Species-Climate Impact Models Under Climate
Change.”</span> <em>Global Change Biology</em> 11: 1504–13.
</div>
<div id="ref-Elith2010" class="csl-entry">
Elith, J., M. Kearney, and S. Phillips. 2010. <span>“The Art of
Modelling Range-Shifting Species.”</span> <em>Methods in Ecology and
Evolution</em> 1: 330–42.
</div>
<div id="ref-Elith2009" class="csl-entry">
Elith, J., and J. R. Leathwick. 2009. <span>“Species Distribution
Models: Ecological Explanation and Prediction Across Space and
Time.”</span> <em>Annual Review of Ecology, Evolution, and
Systematics</em> 40: 677–97.
</div>
<div id="ref-Franklin2010" class="csl-entry">
Franklin, J. 2010. <em>Mapping Species Distributions: Spatial Inference
and Prediction</em>. Cambride University Press.
</div>
<div id="ref-Gillings2019" class="csl-entry">
Gillings, S., D. E. Balmer, B. J. Caffrey, I. S. Downie, D. W. Gibbons,
P. C. Lack, J. B. Reid, J. T. R. Sharrock, R. L. Swann, and R. J.
Fuller. 2019. <span>“Breeding and Wintering Bird Distributions in
Britain and Ireland from Citizen Science Bird Atlases.”</span>
<em>Global Ecology and Biogeography</em> 28 (7): 866–74. <a
href="https://doi.org/10.1111/geb.12906">https://doi.org/10.1111/geb.12906</a>.
</div>
<div id="ref-guisan2005" class="csl-entry">
Guisan, A., and W. Thuiller. 2005. <span>“Predicting Species
Distribution: Offering More Than Simple Habitat Models.”</span>
<em>Ecology Letters</em> 8: 993–1009.
</div>
<div id="ref-Guisan2017" class="csl-entry">
Guisan, A., W. Thuiller, and N. E. Zimmermann. 2017. <em>Habitat
Suitability and Distribution Models with Applications in r</em>.
Cambride University Press.
</div>
<div id="ref-Guisan2000" class="csl-entry">
Guisan, A., and N. E. Zimmermann. 2000. <span>“Predictive Habitat
Distribution Models in Ecology.”</span> <em>Ecological Modelling</em>
135: 147–86.
</div>
<div id="ref-Liu2005" class="csl-entry">
Liu, C., P. M. Berry, T. P. Dawson, and R. G. Pearson. 2005.
<span>“Selecting Thresholds of Occurrence in the Prediction of Species
Distributions.”</span> <em>Ecography</em> 28: 385–93.
</div>
<div id="ref-Roberts2017" class="csl-entry">
Roberts, D. R., V. Bahn, S. Ciuti, M. S. Boyce, J. Elith, G.
Guillera-Arroita, S: Hauenstein, et al. 2017. <span>“Cross-Validation
Strategies for Data with Temporal, Spatial, Hierarchical, or
Phylogenetic Structure.”</span> <em>Ecography</em> 40: 913–29.
</div>
<div id="ref-Zurell2012" class="csl-entry">
Zurell, D., J. Elith, and B. Schroeder. 2012. <span>“Predicting to New
Environments: Tools for Visualising Model Behaviour and Impacts on
Mapped Distributions.”</span> <em>Diversity and Distributions</em> 18:
628–34.
</div>
<div id="ref-Zurell2020a" class="csl-entry">
Zurell, D., J. Franklin, C. König, P. J. Bouchet, C. F. Dormann, J.
Elith, G. Fandos, et al. 2020. <span>“A Standard Protocol for Reporting
Species Distribution Models.”</span> <em>Ecography</em> 43 (9): 1261–77.
<a
href="https://doi.org/10.1111/ecog.04960">https://doi.org/10.1111/ecog.04960</a>.
</div>
</div>
</div>

<!DOCTYPE html>
<html>

<br>
<hr />
<div id="footer">
<p>Damaris Zurell 2022 <a href="http://creativecommons.org/licenses/by/4.0/" >(CC BY 4.0)</a>.  </p>
</div>

</html>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
